components/camel-casper/src/test/java/org/apache/camel/component/casper/consumer/sse/DataSetController.java

```diff
@@ -0,0 +1,144 @@
+	ObjectMapper objectMapper = new ObjectMapper();
+
+	public DataSetController(DataSetService dataSetService) {
+		this.dataSetService = dataSetService;
+	}
+
+	@GetMapping("/events/main")
+	public SseEmitter fetchmain() {
+		SseEmitter emitter = new SseEmitter(6000l);
+
+		ExecutorService executor = Executors.newSingleThreadExecutor();
+		executor.execute(() -> {
+			List<BlockData> dataSets = dataSetService.getBlocks();
+			try {
+
+				// emit added blocks events
+				for (BlockData dataSet : dataSets) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(dataSet));
+				}
+
+				// emit processed deploys events
+				List<DeployProcessedData> datas = dataSetService.getProcessedDeploys();
+				for (DeployProcessedData dat : datas) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(dat));
+
+				}
+
+				// emit steps events
+				List<StepData> steps = dataSetService.getSteps();
+				for (StepData step : steps) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(step));
+				}
+
+				// emit fault events
+				List<FaultData> faults = dataSetService.getFaults();
+				for (FaultData fault : faults) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(fault));
+				}
+
+				List<DeployExpiredData> expireds = dataSetService.getDeploysExpired();
+				// emit expired deploys events
+				for (DeployExpiredData exp : expireds) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(exp));
+				}
+
+				emitter.complete();
+			} catch (IOException e) {
+				emitter.completeWithError(e);
+			}
+		});
+		executor.shutdown();
+		return emitter;
+	}
+
+	@GetMapping("/events/deploys")
+	public SseEmitter fetchdeploys() {
+		SseEmitter emitter = new SseEmitter(5000l);
+
+		ExecutorService executor = Executors.newSingleThreadExecutor();
+		executor.execute(() -> {
+			List<DeployAcceptedData> dataSets = dataSetService.getAcceptedDeploys();
+			try {
+
+				// emit accepetd deploys events
+				for (DeployAcceptedData dataSet : dataSets) {
+					randomDelay();
+					emitter.send(objectMapper.writeValueAsString(dataSet));
+				}
+
+				emitter.complete();
+			} catch (IOException e) {
+				emitter.completeWithError(e);
+			}
+		});
+		executor.shutdown();
+
+		return emitter;
+	}
+
+	@GetMapping("/events/sigs")
+	public SseEmitter fetchsigs() {
+		SseEmitter emitter = new SseEmitter(5000l);
+		// emit finality signatures events
+		ExecutorService executor = Executors.newSingleThreadExecutor();
+		executor.execute(() -> {
+			List<FinalitySignatureData> dataSets = dataSetService.getFinalitySignatures();
+			try {
+				for (FinalitySignatureData dataSet : dataSets) {
+					randomDelay();
+					emitter.send(dataSet);
+				}
+
+				emitter.complete();
+			} catch (IOException e) {
+				emitter.completeWithError(e);
+			}
+		});
+		executor.shutdown();
+		return emitter;
+	}
+
+	private void randomDelay() {
+		try {
+			Thread.sleep(0);

```

Comment:
Please avoid using `Thread.sleep` as much as possible. Instead, please try to replace the code with Awaitility or, in case of synchronization / concurrency, latches and other concurrency handling mechanisms.